File: lib/ts/lte_security_ts33401.c
#include "lte_security_ts33401.h"
#include <stdio.h>

void main()
{
	uint8_t i = 2;
	uint8_t *k_rrc_int;
	uint8_t *k_enb;

	*k_enb = 0x05;
	k_rrc_int = &i;

	
	generate_rrc_integrity_key(k_enb, ID_128_EIA2, k_rrc_int);

	printf("k_rrc_int[0] 0x%x\n", k_rrc_int[0]);
	printf("k_rrc_int[1] %d\n", k_rrc_int[1]);
	printf("k_rrc_int[2] 0x%x\n", k_rrc_int[2]);
	printf("k_rrc_int[3] 0x%x\n", k_rrc_int[3]);
	printf("k_rrc_int[6] 0x%x\n", k_rrc_int[6]);
	
}


/* TS 33.401: A.7 Algorithm key derivation functions

  TS 33.220: B.2 Generic key derivation function 
   FC - single octet used to distinguish between different instances of the algorithm
   P0 ... Pn are the n+1 input parameter encodings
   L0 ... Ln are the two-octet representations of the length of the corresponding input parameter encodings P0..Pn. 

   The final output, i.e. the derived key is equal to the KDF computed on the string S using the key, denoted Key.
   The present document defines the following KDF:
       derived key = HMAC-SHA-256 ( Key , S )
     as specified in [22] and [23]

*/

void generate_rrc_integrity_key(uint8_t *K_eNB, 
	                            enum eps_integrity_algorithm_id algo_id, 
	                            uint8_t *K_RRC_int )
{
	uint8_t FC = 0x15; 
	uint8_t P0 = RRC_int_alg;   /* algorithm type distinguisher */
	uint16_t L0 = 0x0001;       /* length of algorithm type distinguisher (i.e 0x00 0x01) */
	uint8_t P1= algo_id;        /* algorithm identity */
	uint16_t L1 = 0x0001;       /* length of algorithm identity (i.e. 0x00 0x01) */

    /* S = FC || P0 || L0 || P1 || L1 || P2 || L2 || P3 || L3 ||... || Pn || Ln */
    uint8_t S[7] = {FC, P0, L0>>2, L0, P1, L1>>2, L1}; 
	

    //*K_RRC_int = FC;
	K_RRC_int = hmac_sha256(K_eNB, S);
	printf("K_RRC_int[0] 0x%x\n", K_RRC_int[0]);
	printf("K_RRC_int[1] %d\n", K_RRC_int[1]);
	printf("K_RRC_int[2] 0x%x\n", K_RRC_int[2]);
	printf("K_RRC_int[3] 0x%x\n", K_RRC_int[3]);
	printf("K_RRC_int[6] 0x%x\n", K_RRC_int[6]);
}

void generate_user_plane_encryption_key(uint8_t *K_eNB, 
                                        enum eps_encryption_algorithm_id algo_id, 
                                        uint8_t *K_UP_enc)
{
	uint8_t FC = 0x15; 
	uint8_t P0 = UP_enc_alg;  /* algorithm type distinguisher */
	uint16_t L0 = 0x0001;     /* length of algorithm type distinguisher (i.e 0x00 0x01) */
	uint8_t P1= algo_id;        /* algorithm identity */
	uint16_t L1 = 0x0001;     /* length of algorithm identity (i.e. 0x00 0x01) */

	uint8_t S[7] = {FC, P0, L0>>2, L0, P1, L1>>2, L1}; 

	K_UP_enc = hmac_sha256(K_eNB, S);	
}

File: lib/ts/lte_security_ts33401.erl
% Copyright (c) 2018 Aigbe Research
% lte_security_ts33401.erl
%
% Compliance: 
%    3GPP TS 33.401 version 12.13.0 Release 12
%    TS 33.220 V12.3.0 

-module(lte_security_ts33401).

% Table A.7-1: Algorithm type distinguishers
-define(NAS_ENC_ALG, <<1:8>>).
-define(NAS_INT_ALG, <<2:8>>).
-define(RRC_ENC_ALG, <<3:8>>).
-define(RRC_INT_ALG, <<4:8>>).
-define(UP_ENC_ALG,  <<5:8>>).
-define(UP_INT_ALG,  <<6:8>>).

% 5.1.4.1 Integrity requirements
% EPS Integrity Algorithm Identity
-define(ID_EIA0, 0).     % Null Integrity Protection algorithm */
-define(ID_128_EIA1, 1). % SNOW 3G based algorithm */
-define(ID_128_EIA2, 2). % AES based algorithm */
-define(ID_128_EIA3, 3). % ZUC based algorithm */

% 5.1.3.1 Ciphering requirements 
% 5.1.3.2 Algorithm Identifier Values 
% EPS Encryption Algorithm Identity
-define(ID_EEA0, 0). 
-define(ID_128_EEA1, 1).
-define(ID_128_EEA2, 2). 
-define(ID_128_EEA3, 3).

% Use the sha256() hash function to compute a hmac code with the hmac() function
hmac_sha256(Key, Data) ->
	crypto:hmac(sha256, Key, Data, 128).

% key derivation function
%   KRRCint, KRRCenc, KUPint, KUPenc
kdf(KeNB, AlgoType, AlgoId) when size(KeNB) == 32 ->
	FC = <<15:8>>, 
	P0 = AlgoType,         % algorithm type distinguisher
	L0 = <<1:16>>,         % length of algorithm type distinguisher (i.e 0x00 0x01)
	P1= AlgoId,            % algorithm identity
	L1 = <<1:16>>,         % length of algorithm identity (i.e. 0x00 0x01)

    % S = FC || P0 || L0 || P1 || L1 || P2 || L2 || P3 || L3 ||... || Pn || Ln */
    S = <<FC/binary, P0/binary, L0/binary, P1/binary, L1/binary>>, 
    hmac_sha256(K_eNB, S).



% A.2 Kasm derivation function
k_asme_df(CK, IK, MCC, MNC) ->
    MCC_String = integer_to_list(MCC),
    MNC_String = integer_to_list(MNC),
    [MCC3, MCC2, MCC1] = MCC_String,
    [MNC3, MNC2, MNC1] = MNC_String, % 2 or 3

    case length(MNC_String) of
    	2 ->
    		[MNC2, MNC1] = MNC_String,
    		MNC3 = "0";
    	3 ->
    		[MNC3, MNC2, MNC1] = MNC_String
    end,
    SN_id = <<(list_to_integer([MCC2])):4, (list_to_integer([MCC1])):4, 
              (list_to_integer([MCC3])):4, (list_to_integer([MNC3])):4,
              (list_to_integer([MNC2])):4, (list_to_integer([MNC1])):4>>,
	FC = <<10:8>,
	P0 = SN_id,
	L0 = <<3:16>>,
	P1 = SQN xor AK,
	L1 = <<6:16>>,
	S = <<FC/binary, P0/binary, L0/binary, P1/binary, L1/binary>>,
	Key = <<CK/binary, IK/binary>>,
	hmac_sha256(Key, S),
	SN_id.
